### PR TITLE
api/client/service: list running services over replicas

### PR DESCRIPTION
**Pull request by stevvooe**
Show the running number of replicas in `docker service ls`:

```
ID            NAME        REPLICAS  IMAGE   COMMAND
2zs4helqu64f  helloworld  20/5      alpine  ping docker.com      
2zs4helqu64f  train       1/5      alpine  ping docker.com
```

cc @icecrime @vieux

### CODE REVIEW

**Comment by stevvooe** [comment]
(I'm really having trouble with my docker build environment right now, so apologize for the build errors).

File: api/client/service/list.go [hunk 1]
```diff
@@ -60,31 +59,48 @@ func runList(dockerCli *client.DockerCli, opts listOptions) error {
...
+
+		tasks, err := client.TaskList(ctx, types.TaskListOptions{Filter: taskFilter})
+		if err != nil {
+			return err
+		}
+
+		running := map[string]int{}
+		for _, task := range tasks {
+			if task.Status.State == "running" {
+				running[task.ID]++
```
Comment (icecrime) [inline api/client/service/list.go]: This doesn't seem right: shouldn't we be using the ID of the service the task belongs to as index?

**Comment by icecrime** [comment]
LGTM

File: api/client/service/list.go [hunk 2]
```diff
@@ -60,31 +59,48 @@ func runList(dockerCli *client.DockerCli, opts listOptions) error {
...
+		}
+
+		tasks, err := client.TaskList(ctx, types.TaskListOptions{Filter: taskFilter})
+		if err != nil {
+			return err
+		}
+
+		running := map[string]int{}
+		for _, task := range tasks {
+			if task.Status.State == "running" {
```
Comment (vieux) [inline api/client/service/list.go]: `if task.Status.State == swarm.TaskStateRunning {`

**Comment by vieux** [comment]
@stevvooe your example says `20/5` should we be worried ?

**Comment by vieux** [comment]
LGTM besides my comment

**Comment by aluzzardi** [comment]
@stevvooe: You have to filter by desired state, otherwise you end up accounting for tasks not supposed to be running anymore.

There's an API filte for that which you can chain with service=.

I think you should filter by desired RUNNING or desired ACCEPTED. Or filter by hand <=RUNNING but you'll get a ton of extra tasks.

@vieux?

**Comment by aluzzardi** [comment]
Err, my bad, just realized how this works - I though we were listing tasks and counting how many were in desired running vs actual running.

This approach works as well but we don't get the count for global services - but maybe that's okay?

In any case, this should probably use an API filter on state to reduce the payload. @aaronlehmann @vieux ?

**Comment by aaronlehmann** [comment]
@aluzzardi: It looks like we would have to change the API to accommodate filtering by observed state. Do you think it's worth it? The only scenario I can think of where this makes a big difference is if someone has set up their cluster to keep a lot of history for each task slot - then there may be a lot of dead tasks that bloat the payload. Even so, we may want to wait to see if this is noticeable to users before optimizing.

**Comment by stevvooe** [comment]
> This approach works as well but we don't get the count for global services - but maybe that's okay?

This is omitted, for now.

**Comment by mlaventure** [comment]
**NOTE**: Technically all the tests passed, janky failed because some shims were not properly killed and the shell is waiting on them.

So this is :green_heart:

**Comment by aluzzardi** [comment]
@stevvooe Can you make sure the extra requests don't get sent if using `-q` (ID only display)? Just making sure large clusters don't completely freeze the CLI.

> @aluzzardi: It looks like we would have to change the API to accommodate filtering by observed state. Do you think it's worth it? 

@aaronlehmann: Seems like a valid use case for a user to filter by observed state (which is what `docker ps` does). Doesn't need to happen now and doesn't even need to be an index (we could filter on the API side), but I think it makes sense

**Comment by stevvooe** [comment]
> Can you make sure the extra requests don't get sent if using -q (ID only display)? Just making sure large clusters don't completely freeze the CLI.
This behavior is only in the "loud" branch: https://github.com/docker/docker/pull/23613/files#diff-aae751262520e6dfaa1380e281f24584R60

**Comment by vieux** [comment]
@stevvooe I agree with @aluzzardi we should probably use a filter

**Comment by vieux** [comment]
as @aaronlehmann pointed out, we don't have filters on `current_state` we can add that later.

We should merge this one.

**Comment by vieux** [comment]
ping @sfsmithcha it will impact the docs

**Comment by vieux** [comment]
ping @icecrime @tonistiigi

**Comment by sfsmithcha** [comment]
@vieux, just want to verify this impacts GA docs, but not RC1.

**Comment by vieux** [comment]
@sfsmithcha it will impact RC2

**Comment by stevvooe** [comment]
> @stevvooe I agree with @aluzzardi we should probably use a filter

Can we do this in a separate PR?

**Comment by vieux** [comment]
@stevvooe yes, and as I said earlier, we don't have this filter implemented anyway.

**Comment by tonistiigi** [comment]
LGTM

**Comment by sfsmithcha** [comment]
@vieux think docs are handled in #23650 PTAL

**Pull request merged**